packages/js/components/src/experimental-tree-control/linked-tree-utils.ts

```diff
@@ -0,0 +1,161 @@
+/**
+ * Internal dependencies
+ */
+import { AugmentedItem, Item, LinkedTree } from './types';
+
+type MemoItems = {
+	[ value: AugmentedItem[ 'value' ] ]: LinkedTree;
+};
+
+const shouldItemBeExpanded = (
+	item: LinkedTree,
+	createValue: string | undefined
+): boolean => {
+	if ( ! createValue || ! item.children?.length ) return false;
+	return item.children.some( ( child ) => {
+		if ( new RegExp( createValue || '', 'ig' ).test( child.data.label ) ) {
+			return true;
+		}
+		return shouldItemBeExpanded( child, createValue );
+	} );
+};
+
+function findChildren(
+	items: AugmentedItem[],
+	memo: MemoItems = {},
+	parent?: AugmentedItem[ 'parent' ],
+	createValue?: string | undefined
+): LinkedTree[] {
+	const children: AugmentedItem[] = [];
+	const others: AugmentedItem[] = [];
+
+	items.forEach( ( item ) => {
+		if ( item.parent === parent ) {
+			children.push( item );
+		} else {
+			others.push( item );
+		}
+		memo[ item.value ] = {
+			parent: undefined,
+			data: item,
+			children: [],
+		};
+	} );
+
+	return children.map( ( child ) => {
+		const linkedTree = memo[ child.value ];
+		linkedTree.parent = child.parent ? memo[ child.parent ] : undefined;
+		linkedTree.children = findChildren(
+			others,
+			memo,
+			child.value,
+			createValue
+		);
+		linkedTree.data.isExpanded =
+			linkedTree.children.length === 0
+				? true
+				: shouldItemBeExpanded( linkedTree, createValue );
+		return linkedTree;
+	} );
+}
+
+export function getLinkedTree(
+	items: Item[],
+	value: string | undefined
+): LinkedTree[] {
+	const augmentedItems = items.map( ( i, index ) => ( {
+		...i,
+		index,
+		isExpanded: false,
+	} ) );
+	return findChildren( augmentedItems, {}, undefined, value );
+}
+
+export function toggleNode(
+	tree: LinkedTree[],
+	number: number,
+	value: boolean
+): LinkedTree[] {
+	return tree.map( ( node ) => {
+		return {
+			...node,
+			children: node.children
+				? toggleNode( node.children, number, value )
+				: node.children,
+			data: {
+				...node.data,
+				isExpanded:
+					node.data.index === number ? value : node.data.isExpanded,
+			},
+			...( node.parent
+				? {
+						parent: {
+							...node.parent,
+							data: {
+								...node.parent.data,
+								isExpanded:
+									node.parent.data.index === number
+										? value
+										: node.parent.data.isExpanded,
+							},
+						},
+				  }
+				: {} ),
```

Comment:
This seems very complex, but it's just verbose. It's updating `isExpanded` for the parent, if it exists.